pkg/storage/google_cloud_storage.go

```diff
@@ -0,0 +1,213 @@
+func (s *GoogleCloudStorage) getObjectHandleWithRetries(ctx context.Context, bucket, name string) (*storage.ObjectHandle, context.Context, context.CancelFunc) {
+	o := s.client.Bucket(bucket).Object(name).Retryer(
+		storage.WithBackoff(gax.Backoff{
+			Initial:    s.cfg.initialRetryDelay,
+			Max:        s.cfg.maxRetryDelay,
+			Multiplier: s.cfg.retryMultiplier,
+		}),
+		storage.WithPolicy(storage.RetryAlways),
+	)
+	ctx, cancel := context.WithTimeout(ctx, s.cfg.retryTimeout)
+	return o, ctx, cancel
+}
+
+func (s *GoogleCloudStorage) UploadObject(ctx context.Context, bucket, name string, contents []byte, contentType string, metadata map[string]string) error {
+	o, ctx, cancel := s.getObjectHandleWithRetries(ctx, bucket, name)
+	defer cancel()
+
+	writer := o.NewWriter(ctx)
+
+	if contentType != "" {
+		writer.ContentType = contentType
+	}
+
+	if metadata != nil {
+		writer.Metadata = metadata
+	}
+
+	if _, err := writer.Write(contents); err != nil {
+		return fmt.Errorf("failed to write data: %w", err)
+	}
+
+	if err := writer.Close(); err != nil {
+		return fmt.Errorf("failed to close writer: %w", err)
+	}
+
+	return nil
+}
+
+func (s *GoogleCloudStorage) GetObject(ctx context.Context, bucket, name string) ([]byte, error) {
+	o, ctx, cancel := s.getObjectHandleWithRetries(ctx, bucket, name)
+	defer cancel()
+
+	r, err := o.NewReader(ctx)
+	if err != nil {
+		if errors.Is(err, storage.ErrObjectNotExist) {
+			return nil, fmt.Errorf("failed to get object: %w", err)
+		}
+
+		return nil, fmt.Errorf("failed to get google cloud storage reader: %w", err)
+	}
+	defer r.Close()
+
+	var data bytes.Buffer
+	if _, err := io.Copy(&data, r); err != nil {
+		return nil, fmt.Errorf("failed to get object: %w", err)
+	}
+
+	return data.Bytes(), nil
+}
+
+func (s *GoogleCloudStorage) GetObjectWithLimit(ctx context.Context, bucket, name string, limit int64) ([]byte, error) {
+	o, ctx, cancel := s.getObjectHandleWithRetries(ctx, bucket, name)
+	defer cancel()
+
+	r, err := o.NewReader(ctx)
+	if err != nil {
+		if errors.Is(err, storage.ErrObjectNotExist) {
+			return nil, fmt.Errorf("failed to get object: %w", err)
+		}
+
+		return nil, fmt.Errorf("failed to get google cloud storage reader: %w", err)
+	}
+	defer r.Close()
+
+	lr := io.LimitReader(r, limit)
+
+	var data bytes.Buffer
+	if _, err := io.Copy(&data, lr); err != nil {
+		return nil, fmt.Errorf("failed to get object: %w", err)
+	}
+
+	return data.Bytes(), nil
+}
+
+func (s *GoogleCloudStorage) GetMetadata(ctx context.Context, bucket, name string) (map[string]string, error) {
+	o, ctx, cancel := s.getObjectHandleWithRetries(ctx, bucket, name)
+	defer cancel()
+
+	attrs, err := o.Attrs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get object metadata: %w", err)
+	}
+	return attrs.Metadata, nil
+}
+
+func (s *GoogleCloudStorage) DeleteObject(ctx context.Context, bucket, name string, ignoreNotFound bool) error {
```

Comment:
ignoreNotFound feels like a weird API; why would we _not_ want to ignore it? The caller says "ensure object FOO is deleted". If it doesn't exist, the invariant is satisfied.